src/simulation/m_time_steppers.fpp

```diff
@@ -892,6 +910,127 @@ contains
+
+        real(kind(0d0)), dimension(num_fluids) :: alpha_rho  !< Cell-avg. partial density
+        real(kind(0d0)) :: rho        !< Cell-avg. density
+        real(kind(0d0)), dimension(num_dims) :: vel        !< Cell-avg. velocity
+        real(kind(0d0)) :: vel_sum    !< Cell-avg. velocity sum
+        real(kind(0d0)) :: pres       !< Cell-avg. pressure
+        real(kind(0d0)), dimension(num_fluids) :: alpha      !< Cell-avg. volume fraction
+        real(kind(0d0)) :: gamma      !< Cell-avg. sp. heat ratio
+        real(kind(0d0)) :: pi_inf     !< Cell-avg. liquid stiffness function
+        real(kind(0d0)) :: qv         !< Cell-avg. fluid reference energy
+        real(kind(0d0)) :: c          !< Cell-avg. sound speed
+        real(kind(0d0)) :: E          !< Cell-avg. energy
+        real(kind(0d0)) :: H          !< Cell-avg. enthalpy
+        real(kind(0d0)), dimension(2) :: Re         !< Cell-avg. Reynolds numbers
+        real(kind(0d0)) :: blkmod1, blkmod2 !<
+            !! Fluid bulk modulus for Woods mixture sound speed
+        type(int_bounds_info) :: ix, iy, iz
+        type(vector_field) :: gm_alpha_qp
+
+        integer :: i, j, k, l, q !< Generic loop iterators
+
+        real(kind(0d0)) :: dt_local
+
+        integer :: Nfq
+        real(kind(0d0)) :: fltr_dtheta   !<
+         !! Modified dtheta accounting for Fourier filtering in azimuthal direction.
+
+        ix%beg = 0; iy%beg = 0; iz%beg = 0
+        ix%end = m; iy%end = n; iz%end = p
+        call s_convert_conservative_to_primitive_variables( &
+            q_cons_ts(1)%vf, &
+            q_prim_vf, &
+            gm_alpha_qp%vf, &
+            ix, iy, iz)
+
+        ! Computing Stability Criteria at Current Time-step ================
+        !$acc parallel loop collapse(3) gang vector default(present) private(alpha_rho, vel, alpha, Re, fltr_dtheta, Nfq)
+        do l = 0, p
+            do k = 0, n
+                do j = 0, m
+
+                    do i = 1, num_fluids
+                        alpha_rho(i) = q_prim_vf(i)%sf(j, k, l)
+                        alpha(i) = q_prim_vf(E_idx + i)%sf(j, k, l)
+                    end do
+
+                    if (bubbles) then
+                        call s_convert_species_to_mixture_variables_bubbles_acc(rho, gamma, pi_inf, qv, alpha, alpha_rho, Re, j, k, l)
+                    else
+                        call s_convert_species_to_mixture_variables_acc(rho, gamma, pi_inf, qv, alpha, alpha_rho, Re, j, k, l)
+                    end if
+
+                    do i = 1, num_dims
+                        vel(i) = q_prim_vf(contxe + i)%sf(j, k, l)
+                    end do
+
+                    vel_sum = 0d0
+                    do i = 1, num_dims
+                        vel_sum = vel_sum + vel(i)**2d0
+                    end do
+
+                    pres = q_prim_vf(E_idx)%sf(j, k, l)
+
+                    E = gamma*pres + pi_inf + 5d-1*rho*vel_sum + qv
+
+                    H = (E + pres)/rho
```

Comment:
this only works for the 5eqn model without bubbles I believe. perhaps you should be using some other already-existing subroutine to get these quantities? (or creating your own that can be used elsewhere, this computation is repeated several times, e.g., in Riemann solvers and m_data_output).

Reply:
These quantities are calculated in the same way when computing the CFL in `s_write_runtime_info`, so I'm not sure if this is correct. I copy and pasted much of this subroutine.

Reply:
if you're copy/pasting then shouldn't the logic be extracted to avoid duplicating code?

Reply:
i also don't trust the CFL number computed in runtime_info. It might be sufficient for an output file, but this seems to avoid some edge cases I mentioned?

Reply:
What would you suggest naming a subroutine that contains these lines? There's no obvious name for what these lines of code do to me.

Reply:
compute_cfl?

Reply:
I interpreted you suggestion as abstracting these lines of code into a subroutine 
```                 
                    do i = 1, num_fluids
                        alpha_rho(i) = q_prim_vf(i)%sf(j, k, l)
                        alpha(i) = q_prim_vf(E_idx + i)%sf(j, k, l)
                    end do

                    if (bubbles) then
                        call s_convert_species_to_mixture_variables_bubbles_acc(rho, gamma, pi_inf, qv, alpha, alpha_rho, Re, j, k, l)
                    else
                        call s_convert_species_to_mixture_variables_acc(rho, gamma, pi_inf, qv, alpha, alpha_rho, Re, j, k, l)
                    end if

                    do i = 1, num_dims
                        vel(i) = q_prim_vf(contxe + i)%sf(j, k, l)
                    end do

                    vel_sum = 0d0
                    do i = 1, num_dims
                        vel_sum = vel_sum + vel(i)**2d0
                    end do

                    pres = q_prim_vf(E_idx)%sf(j, k, l)

                    E = gamma*pres + pi_inf + 5d-1*rho*vel_sum + qv

                    H = (E + pres)/rho
```

Reply:
`compute_cfl_from_dt` and `compute_dt_from_cfl` would both call whatever these lines are abstracted into

Reply:
ultimately, computing the CFL (or dt from CFL) requires the speed of sound, the mesh spacing, and the time step size. so we need a routine that can get the speed of sound from whatever variables we have at present, regardless of multiphase model. this should already be encapsulated in the `s_compute_speed_of_sound` routine if I understand correctly. If it isn't, then can you explain to me what's going on

Reply:
but you are (obviously) correct tht compute_cfl_from_dt and the reverse would call some of these lines of code, but it looks like we are doing this in multiple places? even if we are only doing it in one place, it still deserves its own routine.

Reply:
`s_compute_speed_of_sound` accounts for the model being used as well as sub-grid bubbles. I'll just throw these into a subroutine call `s_get_cell_state`.

Reply:
it looks like in the end you are trying to get `H`, which is actually repeated code from the Riemann solvers as well. so we should be able to use the same code in several places.

Reply:
`s_compute_enthalpy`, I suppose. similarly, we have a subroutine called `s_compute_pressure` (or some such) that is used in a similar way.

Reply:
Splitting it into `s_compute_pressure` and `s_compute_enthalpy` makes sense. Though `s_compute_enthalpy` will be a 1 line subroutine.

Reply:
I think `s_compute_mixture_variables` actually makes the most sense, since this is what all these lines ultimately do.

Reply:
Well, there's already a subroutine called that. I guess I'll do `s_compute_cell_state`

Reply:
but it isnt the cell state per se, you are just computing the enthalpy, right? perhaps you can create a function called `compute_enthalpy` that can all other routines as above, including the `compute_mixture_variables`